grove/tomography/utils.py

```diff
@@ -0,0 +1,640 @@
+
+THREE_COLOR_MAP = ['#48737F', '#FFFFFF', '#D6619E']
+rigetti_3_color_cm = LinearSegmentedColormap.from_list("Rigetti", THREE_COLOR_MAP[::-1], N=100)
+
+FIVE_COLOR_MAP = ['#C671A2', '#545253', '#85B5BE', '#ECE9CC', '#C671A2']
+rigetti_4_color_cm = LinearSegmentedColormap.from_list("Rigetti", FIVE_COLOR_MAP[::-1], N=100)
+
+FROBENIUS = 'fro'
+CHOI = 'choi'
+
+
+QX = qt.sigmax()
+QY = qt.sigmay()
+QZ = qt.sigmaz()
+QI = qt.qeye(2)
+
+EPS = 1e-8
+
+
+def sample_outcomes(probs, n):
+    """
+    For a discrete probability distribution ``probs`` with outcomes 0, 1, ..., k-1 draw ``n``
+    random samples.
+
+    :param list probs: A list of probabilities.
+    :param Number n: The number of random samples to draw.
+    :return: An array of samples drawn from distribution probs over 0, ..., len(probs) - 1
+    :rtype: numpy.ndarray
+    """
+    dist = np.cumsum(probs)
+    rs = np.random.rand(n)
+    return np.array([(np.where(r < dist)[0][0]) for r in rs])
+
+
+def basis_state_preps(*qubits):
+    """
+    Generate a sequence of programs that prepares the measurement
+    basis states of some set of qubits in the order such that the qubit
+    with highest index is iterated over the most quickly:
+    E.g., for qubits=(0, 1), it returns the circuits::
+
+        I_0 I_1
+        I_0 X_1
+        X_0 I_1
+        X_0 X_1
+    :param list qubits: Each qubit to include in the basis state preparation.
+    :return: Yields programs for each basis state preparation.
+    :rtype: Program
+    """
+    for prep in cartesian_product([I, X], repeat=len(qubits)):
+        basis_prep = Program()
+        for gate, qubit in zip(prep, qubits):
+            basis_prep.inst(gate(qubit))
+        yield basis_prep
+
+
+def basis_labels(n):
+    """
+    Generate a list of basis labels for `n` qubits, ordered from least to greatest, in big-endian
+     format:
+
+        ['00..00', '00..01', ..., '11..11']
+
+    :param n:
+    :return: A list of strings of length n that enumerate the n-qubit bitstrings
+    :rtype: list
+    """
+    return ["".join(labels) for labels in itertools.product('01', repeat=n)]
+
+
+def sample_bad_readout(program, num_samples, assignment_probs, cxn):
+    """
+    Generate `n` samples of measuring all outcomes of a Quil `program`
+    assuming the assignment probabilities `assignment_probs` by simulating the
+    wave function on a reference-qvm SyncConnection `cxn`
+
+    :param pyquil.quil.Program program: The program.
+    :param int num_samples: The number of samples
+    :param numpy.ndarray assignment_probs: A matrix of assignment probabilities
+    :param reference_qvm.api.SyncConnection cxn: the QVM connection.
+    :return: The resulting sampled outcomes from assignment_probs applied to cxn, one dimensional.
+    :rtype: numpy.ndarray
+    """
+    wf = cxn.wavefunction(program)
+    return sample_outcomes(assignment_probs.dot(abs(wf.amplitudes.ravel())**2), num_samples)
+
+
+def make_histogram(samples, ksup):
```

Comment:
he ... I indicated this above. I feel this is an unneccessary layer we can remove?

Reply:
Between the tests and tomography we use this call around ten times - writing the `arange(ksup + 1) -.5)[0]` logic that many times feels redundant. I'd rather not remove it, but if you think it's better to copy the code that many times I'll do it.

Reply:
Yeah, and it may not be obvious to users why one has to provide these fixed bins. I like abstracting/packaging the call

Reply:
👍